API.md

```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
 
 `{"error":{"id":"no_login","message":"You are not logged in to an account"},"status":"error"}`
 
+### `/resetpass` - Reset a password for a user
+#### Params
+- `account`
+- `email`
+
+#### Responses
+##### Success
+
+`{"status":"ok"}`
+
+#### Errors
+
+`{"error":{"id":"no_account","message":"Unable to find matching account"},"status":"error"}`
+
+`{"error":{"id":"mail_failed","message":"Unable to send reset email"},"status":"error"}`
+
+### `/resetpass/confirm` - Confirm a password reset
+This has two main uses:
+- Simply confirming the reset request and logging in the user, which will then mark the account as unconfirmed until the user changes their password, or
+- Confirming the reset request and also setting the password to `newpass` in one request (requires sending the `newpass` parameter
+
+This allows finer control from the API consumer side as to the separation of authentication and modification of the account
+
+#### Params
+- `account` - Account name to reset the password for
+- `code` - Confirmation code from the user
+- `newpass` _optional_ - If set, sets the new password in this request, otherwise a later request to `/user/set/password` is required.
+
+#### Responses
+##### Success
+
+`{"status":"ok","password_set":true|false,"account":"<ACCOUNT_NAME>","verified":true|false,"session":"<SESSION_ID>"}`
+
+- `password_set` - Was the password changed in this request or is a future request to `/user/set/password` required
+- `verified` - Is the account a verified account or not
+
+##### Errors
+
+`{"error":{"id":"no_account","message":"Unable to find matching account"},"status":"error"}`
+
+`{"error":{"id":"wrong_code","message":"Invalid reset token"},"status":"error"}`
+
+`{"error":{"id":"expired_code","message":"Expired reset token"},"status":"error"}`
+
+`{"error":{"id":"invalid_password","message":"That password is invalid"},"status":"error"}`
+
+### `/user/set/password` - Set the user's password
+#### Params
+- `session`
+- `newpass`
+
+#### Responses
+##### Success
+
+`{"status":"ok"}`
+
+##### Errors
+
+`{"error":{"id":"no_login","message":"You are not logged in to an account"},"status":"error"}`
+
+`{"error":{"id":"invalid_password","message":"That password is invalid"},"status":"error"}`
```

Comment:
see above regarding exposing information about existing accounts. Unless this only hits users that are logged in. In which case further information about why their selected password is invalid may be prudent.

Reply:
This is simply about the provided new password, and currently there isn't an easy way to tell why it was considered invalid, it behaves the same as the registration endpoint

Reply:
This is an internal API only, it doesn't matter if it exposes if the account is available or not here as it's only snoonet/irc.com that uses it

Reply:
> This is an internal API only, it doesn't matter if it exposes if the account is available or not here as it's only snoonet/irc.com that uses it

That does not change the fact that internal APIs should still not blindly trust their counterparts. That's how internal breaches allow for lateral moves and additional access to things other than the breached component. Security and layers thereof should still exist in internal environments.

>This is simply about the provided new password, and currently there isn't an easy way to tell why it was considered invalid, it behaves the same as the registration endpoint

Thats fine. My concern was regarding exposing information to a connection that was not authed for the account its attempting to mess with.

Reply:
> Thats fine. My concern was regarding exposing information to a connection that was not authed for the account its attempting to mess with.

Ah, yeah that's not happening here, this endpoint requires a valid and authenticated session token